File: backend/src/ra_agent/api/approvals.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, Request

from ra_agent.contracts import APIResponse, ApprovalDecision, ApprovalStatus
from ra_agent.core.container import ServiceContainer

from .deps import get_services
from .task_graphs import is_known_graph_task, list_approval_views

router = APIRouter(prefix="/api/approvals", tags=["approvals"])

# Phase 4 production integration must supply this trusted caller identity from auth.


@router.get("")
async def list_approvals(
    request: Request,
    services: Annotated[ServiceContainer, Depends(get_services)],
    status: ApprovalStatus | None = None,
    task_id: str | None = None,
) -> APIResponse[list[dict[str, str]]]:
    if task_id is not None and not is_known_graph_task(request, task_id) and task_id not in getattr(
        request.app.state, "task_runs", {}
    ):
        raise HTTPException(status_code=404, detail="Unknown task")
    return APIResponse(data=await list_approval_views(services, task_id=task_id, status=status))


@router.post("/{approval_id}/grant")
async def grant(
    approval_id: str,
    services: Annotated[ServiceContainer, Depends(get_services)],
    decided_by: Annotated[str, Query(min_length=1, pattern=r".*\S.*")],
    reason: str = "approved",
) -> APIResponse[ApprovalDecision]:
    try:
        decision = await services.approval_service.grant(
            approval_id, decided_by, reason
        )
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    return APIResponse(data=decision)


@router.post("/{approval_id}/deny")
async def deny(
    approval_id: str,
    services: Annotated[ServiceContainer, Depends(get_services)],
    decided_by: Annotated[str, Query(min_length=1, pattern=r".*\S.*")],
    reason: str = "denied",
) -> APIResponse[ApprovalDecision]:
    try:
        decision = await services.approval_service.deny(
            approval_id, decided_by, reason
        )
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    return APIResponse(data=decision)

File: frontend/src/api/approvals.ts
import { get, post } from './client';

export interface ApprovalItem {
  approval_id: string;
  status: string;
  tool_name: string;
  reason: string;
  step_id: string;
  request_id: string;
  task_id: string;
}

export function listApprovals(params?: {
  status?: string;
  task_id?: string;
}): Promise<ApprovalItem[]> {
  const query = new URLSearchParams();
  if (params?.status) query.set('status', params.status);
  if (params?.task_id) query.set('task_id', params.task_id);
  const qs = query.toString();
  return get(`/approvals${qs ? '?' + qs : ''}`);
}

export function grantApproval(
  approvalId: string,
  decidedBy: string,
  reason?: string,
): Promise<{ approval_id: string; status: string; decided_by: string; reason: string }> {
  const qs = new URLSearchParams({ decided_by: decidedBy });
  if (reason) qs.set('reason', reason);
  return post(`/approvals/${approvalId}/grant?${qs.toString()}`);
}

export function denyApproval(
  approvalId: string,
  decidedBy: string,
  reason?: string,
): Promise<{ approval_id: string; status: string; decided_by: string; reason: string }> {
  const qs = new URLSearchParams({ decided_by: decidedBy });
  if (reason) qs.set('reason', reason);
  return post(`/approvals/${approvalId}/deny?${qs.toString()}`);
}
